### PR TITLE
[AutoDiff] Introduce `autodiff_function_extract` instruction

### PR DESCRIPTION
**Pull request by rxwei**
### Introduction

An `autodiff_function_extract` instruction extracts an associated differentiation function from the given ``@autodiff`` function at a specific differentiation order. It must be provided with an associated function kind: ``[jvp]`` or ``[vjp]``.

`autodiff_function_extract` pairs with `autodiff_function` introduced in #20505.

### Syntax

```ebnf
  sil-instruction ::= 'autodiff_function_extract'
                      sil-autodiff-associated-function-kind
                      sil-autodiff-function-differentiation-order
                      sil-value ':' sil-type

  sil-autodiff-associated-function-kind ::= '[' sil-autodiff-associated-function-kind-name ']'
  sil-autodiff-associated-function-kind-name ::= 'jvp' | 'vjp'
  sil-autodiff-function-differentiation-order ::= '[' 'order' [0-9]+ ']'
```

### Examples

```swift
  autodiff_function_extract [vjp] [order 1] %0 : $@autodiff (T) -> T
  autodiff_function_extract [jvp] [order 1] %0 : $@autodiff (T) -> T
```

### Other changes

- Drop the differential and pullback functions from the `@autodiff` function bundle layout, because they are no longer necessary.
- Drop "legacy reverse mode" from `autodiff_function` instruction.
- Include a utility to compute the JVP/VJP type: `SILFunctionType::getAutoDiffAssociatedFunctionType`.

### CODE REVIEW

**Review by dan-zheng — APPROVED** [review]
LGTM! Design of `autodiff_function` and `autodiff_function_extract` makes sense.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
cc @bgogul for AD ramp-up.

**Comment by rxwei** [comment]
The macOS failure has been fixed by #20595.

File: include/swift/AST/AutoDiff.h [hunk 1]
```diff
@@ -251,54 +249,36 @@ struct SILAutoDiffConfig {
...
-  Differential,
-  // The Jacobian-vector products operator.
-  VJP,
-  // The pullback, the result of JVP except that it has boxed arguments
-  // (closure captures) for partial application within VJP.
-  Pullback
+/// `autodiff_function_extract` instructions in SIL.
+struct AutoDiffAssociatedFunctionKind {
+  enum innerty : uint8_t {
+     // The vector-Jacobian products operator.
```
Comment (marcrasi) [inline include/swift/AST/AutoDiff.h]: Are the comments correct? The words are in the opposite order of the letters of the enum values.

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -4077,6 +4077,12 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
   CanSILFunctionType getWithDifferentiability(
       unsigned differentiationOrder, const SmallBitVector &parameterIndices);
 
+  /// Returns the differentiation function type associated functions with this
+  /// function type.
```
Comment (marcrasi) [inline include/swift/AST/Types.h]: This comment looks a bit typoed.

**Pull request merged**